include/deal.II/meshworker/copy_data.h

```diff
@@ -0,0 +1,144 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_meshworker_copy_data_h
+#define dealii_meshworker_copy_data_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/types.h>
+
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/vector.h>
+
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace MeshWorker
+{
+  /**
+   * Helper copy data struct.
+   *
+   * This class is a good default drop in CopyData object for the
+   * WorkStream::run() and MeshWorker::mesh_loop() functions.
+   *
+   * It allows you to store arbitrary number of (local) FullMatrices, vectors,
+   * and local_dof_indices vectors.
+   *
+   * @author Luca Heltai, 2019.
+   */
+  template <int n_matrices    = 1,
+            int n_vectors     = n_matrices,
+            int n_dof_indices = n_matrices>
+  struct CopyData
+  {
+    /**
+     * Initialize everything with the same @p size.
+     */
+    CopyData(const unsigned int size);
+
+    /**
+     * For every object, specify the size they should have.
+     */
+    CopyData(
+      const std::array<std::array<unsigned int, 2>, n_matrices> &matrix_sizes,
+      const std::array<unsigned int, n_vectors> &                vector_sizes,
+      const std::array<unsigned int, n_dof_indices> &dof_indices_sizes);
+
+    /**
+     * Deep copy constructor.
+     */
+    CopyData(const CopyData<n_matrices, n_vectors, n_dof_indices> &other) =
+      default;
+
+    /**
+     * Allow resetting of all elements of the struct to zero, by simply
+     * calling (*this) = 0;
+     */
+    void
+    operator=(const double &number);
+
+    /**
+     * Array of local matrices.
+     */
+    std::array<FullMatrix<double>, n_matrices> matrices;
+
+    /**
+     * Array of local vectors.
+     */
+    std::array<Vector<double>, n_vectors> vectors;
+
+    /**
+     * Array of local degrees of freedom indices.
+     */
+    std::array<std::vector<types::global_dof_index>, n_dof_indices>
+      local_dof_indices;
+  };
+
+
+
+  //
+  // Template definitions
+  //
+  template <int n_matrices, int n_vectors, int n_dof_indices>
+  CopyData<n_matrices, n_vectors, n_dof_indices>::CopyData(
+    const unsigned int size)
+  {
+    for (auto &m : matrices)
+      m.reinit({size, size});
```

Comment:
Should we first check if each of these reinits and resizing operations are necessary? e.g.
```c++
    for (auto &m : matrices)
      if (m.m() != size || m.n() != size)
        m.reinit({size, size});
```
Same below and in next function.

Reply:
This check is already done by the reinit function itself.

Reply:
Ok, thanks for the clarification!

Reply:
Also, you're in the constructor -- the initial size of the matrices is for sure zero :-)